Make MBEDTLS_CHECK_PARAMS usable without defining mbedtls_param_failed

gilles-peskine-arm opened the pull request:
Tweak `config.h` and `platform_util.h` so that `config.pl full` also enables the definition of `MBEDTLS_PARAM_FAILED` as `assert`. Adapt `all.sh` so that invalid-parameter tests are executed, but programs don't need to provide an `mbedtls_param_failed` function. Fix https://github.com/ARMmbed/mbedtls/issues/2671

I also included two small commits:
* When building with `-m32`, also link with `-m32`, as discussed in https://github.com/ARMmbed/mbedtls/pull/1622#discussion_r290718810
* Fix #2682 `programs` getting rebuilt over and over.

The primary goal of this PR is to unblock https://github.com/ARMmbed/mbedtls/pull/1622.

This PR requires the crypto submodule to have the changes from https://github.com/ARMmbed/mbed-crypto/pull/148. To pass CI, I include a commit that updates the submodule, which should be removed once the precursor PR has been merged into the submodule.

Backports:

* 2.16: yes, to keep CI aligned, except for the makefile fix which is not applicable. → #2701
* 2.7: yes, but only the `-m32` link option, everything else is not applicable. → #2702

Patater commented on tests/scripts/all.sh:
```diff
@@ -832,9 +832,21 @@ component_test_no_use_psa_crypto_full_cmake_asan() {
     if_build_succeeded env OPENSSL_CMD="$OPENSSL_NEXT" tests/compat.sh -e '^$' -f 'ARIA\|CHACHA'
 }
```
In the commit message of commit "Add all.sh component that exercises invalid_param checks ", this sentence doesn't make sense.

"Add a component to all.sh that enables MBEDTLS_CHECK_PARAMS but disables MBEDTLS_CHECK_PARAMS_ASSERT doesn't define MBEDTLS_PARAM_FAILED so that these tests are exercised."

Should this be like so? (I only added an "and".)

"Add a component to all.sh that enables MBEDTLS_CHECK_PARAMS but disables MBEDTLS_CHECK_PARAMS_ASSERT and doesn't define MBEDTLS_PARAM_FAILED so that these tests are exercised."

gilles-peskine-arm replied:
I fixed this and reworded the commit message to make it easier to read.

Patater commented on tests/scripts/all.sh:
```diff
@@ -832,9 +832,21 @@ component_test_no_use_psa_crypto_full_cmake_asan() {
     if_build_succeeded env OPENSSL_CMD="$OPENSSL_NEXT" tests/compat.sh -e '^$' -f 'ARIA\|CHACHA'
 }
 
+component_test_check_params_tests () {
```
Suggest a more descriptive name, like `component_test_check_mbedtls_param_failed()`. It's a bit verbose if you also add "tests" into the name to describe that this only runs the tests. I think it more important to emphasize that it is testing the `mbedtls_param_failed()` function, more so than that this runs only the tests.

gilles-peskine-arm replied:
It's the only component that runs the tests that verify that calls to API functions that pass invalid parameters result in a call to `MBEDTLS_PARAM_FAILED`. I renamed it to `test_check_params_functionality`.

gilles-peskine-arm commented:
I removed the commit with the crypto submodule change. Expect failures on the pr-head job but the pr-merge job should succeed.

I also modified the `all.sh` commit.

Patater commented:
We still need the commit to update the submodule in this PR (or some other one). It doesn't need to point at a development branch, but should point at the current head of crypto/development.

gilles-peskine-arm commented:
Oh right. Submodule updated.

Patater commented:
ARMmbed/mbed-crypto#148 has been merged. Removing label "needs: preceding PR".

RonEld commented:
This should also fix https://github.com/ARMmbed/mbed-os/issues/9193
